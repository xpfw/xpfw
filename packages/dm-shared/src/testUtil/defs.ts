import { StatType } from "@xpfw/dm"

const simpleSum = {
  type: StatType.sum,
  id: "simpleSum"
}
const pathSum = {
  type: StatType.sum,
  options: {itemPath: "myNum"},
  id: "pathSum"
}

const timeStepSum = {
  type: StatType.timeStep,
  id: "timeStepSum",
  options: {
    subType: StatType.sum
  }
}

const timeStepSumPath = {
  type: StatType.timeStep,
  id: "timeStepSumPath",
  options: {
    subType: StatType.sum,
    subConfig: {itemPath: "myNum"}
  }
}

const timeStepTimeDistance = {
  type: StatType.timeStep,
  id: "timeStepTimeDistance",
  options: {
    subType: StatType.avgPrevTimeDistance,
    subConfig: {itemPath: "createdAt"
  }}
}

export {
  simpleSum, pathSum,
  timeStepSum, timeStepSumPath, timeStepTimeDistance,
  timeStepForm
}
