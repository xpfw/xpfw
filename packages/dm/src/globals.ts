const StatType = {
  sum: "sum",
  mean: "mean",
  timeStep: "timeStep",
  sumTimeDistance: "sumTimeDistance",
  sumPrevTimeDistance: "sumPrevTimeDistance",
  avgTimeDistance: "avgTimeDistance",
  avgPrevTimeDistance: "avgPrevTimeDistance"
}

export {
  StatType
}
