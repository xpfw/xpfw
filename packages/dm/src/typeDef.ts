export interface IStatConfig {
  type: string
  /**
   * used for recognition in react elements
   */
  id: string
  options?: any
}

export type IFindMethod = any

/**
 * Based StatCalculator Type
 * O is the type of the iterated Object
 * R is the type of the result of the calculator
 */
export type TypedStatCalculator<O, R> =
  (object: O) => Promise<R>

/**
 * Common Options for calculators
 */
export interface ICommonOptions {
  itemPath?: string
}

/**
 * Common Options for comparers
 * Will use itemPath and compareWithPath
 */
export interface ICompareOptions extends ICommonOptions {
  /**
   * path for compare with
   */
  compareWithPath?: string
}

/**
 * Common Options for timestep functions
 */
export interface ITimestepOptions extends ICommonOptions {
  /**
   * path for date to check
   */
  datePath?: string
  subConfig?: any
  subType?: string
  timeSteps: any
}

export interface IChartConfig {
  id: string
  calcType: string
  displayType: string
  options?: any
}

export type NumberCalculator = (config?: ICommonOptions,
                                findMethod?: (c: string, q: any) => Promise<any>) => TypedStatCalculator<any, number>

export type CompareNumberCalculator = (config?: ICompareOptions,
                                       findMethod?: (c: string, q: any) => Promise<any>)
                                       => TypedStatCalculator<any, number>

export type TimeStepCalculator = (config?: ITimestepOptions,
                                  findMethod?: (c: string, q: any) => Promise<any>)
                                  => TypedStatCalculator<any, number>

export type DefaultStatCalculator = TypedStatCalculator<any, void>
