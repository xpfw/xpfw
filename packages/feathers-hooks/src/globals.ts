const Method = {
  Create: "create",
  Update: "update",
  Patch: "patch",
  Find: "find",
  Remove: "remove",
  Get: "get"
}

export {
  Method
}
