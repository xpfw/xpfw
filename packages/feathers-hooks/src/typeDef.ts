import { Hook } from "@feathersjs/feathers"
import { ExtendedJSONSchema } from "@xpfw/form"
import { IParameters, IPermissionSchema } from "@xpfw/permission"

export type ValidateHook = (form: ExtendedJSONSchema, custom?: IParameters) => Promise<Hook>
export type GeneralValidateHook = (form: ExtendedJSONSchema, method: string, custom?: IParameters) => Promise<Hook>

export type PermissionHook = (form: IPermissionSchema, custom?: IParameters) => Hook
export type GeneralPermissionHook = (form: IPermissionSchema, method: string, custom?: IParameters) => Hook
