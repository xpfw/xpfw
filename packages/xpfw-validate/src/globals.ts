const RequiredType = {
  Always: 2,
  AllNotSet: 3,
  SetToVal: 4
}

const ErrorType = {
  invalidType: 3,
  required: 4,
  tooLow: 5,
  tooHigh: 420,
  notEqual: 6,
  emptyUpdate: 7
}

const FieldType = {
  Text: 2,
  Number: 3,
  Boolean: 4,
  Date: 5,
  Select: 6,
  Password: 7,
  Array: 10,
  Object: 12,
  RelationshipSingle: 13,
  RelationshipMulti: 14,
  Slider: 15,
  Location: 16
}

const Permission = {
    Guest: "guest",
    User: "user",
    Owner: "owner",
    Admin: "admin",
    Server: "server"
}

const Method = {
  Create: "create",
  Update: "update",
  Patch: "patch",
  Find: "find",
  Remove: "remove",
  Get: "get"
}

const StatType = {
  sum: "sum",
  mean: "mean",
  timeStep: "timeStep",
  sumTimeDistance: "sumTimeDistance",
  sumPrevTimeDistance: "sumPrevTimeDistance",
  avgTimeDistance: "avgTimeDistance",
  avgPrevTimeDistance: "avgPrevTimeDistance"
}

const DateType = {
  dateTime: 2,
  date: 3,
  time: 4
}

const options = {
  userCollection: "users",
  userIdPath: "_id"
}

export {
  StatType, RequiredType, ErrorType, FieldType, Permission, Method, DateType, options
}
export default {
  StatType, RequiredType, ErrorType, FieldType, Permission, Method, DateType, options
}
