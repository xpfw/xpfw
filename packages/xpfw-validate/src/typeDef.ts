export interface ISelectItem {
  label: string
  value: any
}

export type ValidateFunction = (value: any, options: IField, additionalParams: IParametersBase) => Promise<any>

export interface IValidateRequiredOptions {
  type: number
  model?: string | string[]
  values?: any
}

export type FieldConverter = (value: any) => any
export type FieldConvertFunc =
FieldConverter | {create?: FieldConverter, update?: FieldConverter, find?: FieldConverter}

export interface IValidateOptions {
  /**
   * Wether the field should be required and under which circumstances
   */
  required?: IValidateRequiredOptions

  /**
   * Minimum Value (for Number) or Length (for string, array, etc.)
   *
   * @type {number}
   * @memberof IValidateOptions
   */
  min?: number

  /**
   * Maximum Value (for Number) or Length (for string, array, etc.)
   *
   * @type {number}
   * @memberof IValidateOptions
   */
  max?: number

  /**
   * Step Size for number fields
   *
   * @type {number}
   * @memberof IValidateOptions
   */
  step?: number
  /**
   * Recursive ValidationOptions allow for defining Objects or Arrays in one Definition
   *
   * @type {IValidateOptions}
   * @memberof IValidateOptions
   */
  validate?: IValidateOptions
  customOpts?: any
  /**
   * Will check if current value equals the given <Field.mapTo> Key
   * Will check IParametersBase.formData at parameters.formData.<Field.mapTo>
   *
   * @type {string[]}
   * @memberof IValidateOptions
   */
  match?: string[]
  /**
   * Enhance your Validation by adding a custom Check Function
   *
   * @type {ValidateFunction}
   * @memberof IValidateOptions
   */
  custom?: ValidateFunction
  /**
   * For Arrays the expected Type of the content of the array
   * For Date it can be one of global.DateType
   * @type {(number | string)}
   * @memberof IValidateOptions
   */
  type?: number | string
  objectDef?: IField[]

  /**
   * Validate the current user has at least this perm level
   * in order to make ALL actions if number or a specific action
   */
  permission?: IPermTypes

  /** If value isn't set then set it to this one */
  defaultValue?: any

  /**
   * Use this to hide fields in specific contexts
   * Hidden of false / 0, shown if true
   */
  hide?: IFieldVisibilityDef
  /**
   * Customize what FormStore.getFormData produces on a per field basis
   * Like hide and permission this can be set globally or per Method
   * @type {FieldConvetFunc}
   * @memberof IValidateOptions
   */
  convert?: FieldConvertFunc
  /**
   * Collection to search in if Field has type RelationshipSingle or RelationshipMany
   */
  relationshipCollection?: string
  /**
   * Path to get name from if has  type RelationshipSingle or RelationshipMany
   */
  relationshipNamePath?: string
  /**
   * Path to get id from if has  type RelationshipSingle or RelationshipMany
   */
  relationshipIdPath?: string
}

export type SelectOptions = ISelectItem[]

export type SelectOptionGetter = (state: any) => SelectOptions

export type SelectOptionsField = SelectOptions | SelectOptionGetter

export interface IFieldVisibilityDef {
  create?: boolean
  update?: boolean
  find?: boolean
  remove?: boolean
}
export interface IField {
  type: string | number
  mapTo: string
  theme?: string
  validate?: IValidateOptions
}

export interface IFormSection {
  fields: IField[]
  name?: string
}

export interface IStatConfig {
  type: string
  /**
   * used for recognition in react elements
   */
  id: string
  options?: any
}

export interface IForm {
  model: string
  sections: IFormSection[]
  collection?: string
  icon?: any
  multiCollection?: string[]
  permissions?: IPermissionsForForm
  /**
   * Configure numerical xpfw-dm using stats
   * See `globals.StatType` for available types
   */
  stats?: IStatConfig[]
  options?: {
    /**
     * will add belongsTo in ui-shared so Permission.Owner can properly function
     */
    addBelongsTo?: boolean
    /**
     * add and display a createdAt date field
     */
    addCreatedAt?: boolean
    /**
     * path to id for various things
     */
    idPath?: string
    /**
     * Function that will return the custom built query
     *
     * @type {Function}
     */
    queryBuilder?: (form: IForm, prefix?: string) => any
    /**
     * The default sort for this collection
     */
    defaultSort?: any
    /**
     * Wether in the related search a $nin query on the idPath field should be included
     * Be sure to convert the strings into ObjectIds on the server side in case you are using MongoDb!
     */
    filterOutById?: boolean
  }
}

export interface IFormInstance extends IForm {
  getFields: IField[]
}

export interface IFieldSelect extends IField {
  selectOptions?: SelectOptionsField
}

export interface IFieldWithOptions<T> extends IField {
  additionalOptions?: T
}

/**
 * The Registry is needed to enable things like a subpage /select/<formModel>.<fieldMapTo>
 * It can then only with the knowledge of the string find out if this field exists
 * and then display or do whatever with it
 *
 * @export
 * @interface IRegistry
 */
export interface IRegistry {
  forms: {[index: string]: IForm}
  fields: {[index: string]: IField}
  validators: {[index: string]: ValidateFunction}
  registerField: (field: IField) => void
  findField: (mapTo: string) => IField | null
  registerForm: (form: IForm) => void
  findForm: (model: string) => IForm
  registerValidator: (name: string, validator: ValidateFunction) => void
  findValidator: (name: string) => ValidateFunction | null
}

/**
 * Parameters are supposed to be a universal way to pass relevant JS-refs
 * to write validation / permission / whatever functions that can try searching
 * offline data first / throw error if not connected / fetch data if missing via custom
 */
export interface IParametersBase {
  /**
   * The Field that you want to check
   *
   * @type {IField}
   * @memberof IParametersBase
   */
  field?: IField
  /**
   * The Form containing your Field
   * This is required for validatePermission checks
   */
  form?: IForm
  /**
   * Type of method so you can allow fields only in certain situations (find but not update etc.)
   *
   * @type {string}
   * @memberof IParametersBase
   */
  method?: string
  /**
   * If you are searching for a Document save it here
   *
   * @type {string}
   * @memberof IParametersBase
   */
  docId?: string
  doc?: any
  /**
   * Helper Variable you can set to indicate wether server requests are currently possible
   *
   * @type {boolean}
   * @memberof IParametersBase
   */
  isOffline?: boolean
  /**
   * Here you can pass any Object that will help you to achieve your task offline
   *
   * @type {*}
   * @memberof IParametersBase
   */
  offlineData?: any
  /**
   * This should contain the current state mapping of your form
   * e.g. {
   *  "LoginForm.username": "im user"
   * }
   * where LoginForm = IForm.model and username = IField.mapTo
   * @type {*}
   * @memberof IParametersBase
   */
  formData?: any
  /**
   * Form Registry to use for validation.
   *
   * @type {IRegistry}
   * @memberof IParametersBase
   */
  registry: IRegistry
  /**
   * validatePermission will check the ID of a User and search for it here via lodash.get
   */
  idPath?: string
  /**
   * Put the currently authenticated user in here for Permission Checks
   */
  currentUser?: any
  /**
   * validatePermission will check if the current userID equals this key of the object
   * given object {belongsTo: "myUserId"} and currentUser has "myUserId" as id this will return Permission.Owner
   */
  directOwnershipPath?: string
  /**
   * The Permissions Hook will try to feathers.get the object via this serviceName
   * The ID this is called with is gotten from hook at `docIdPath`
   */
  serviceName?: string
  /**
   * See `servicename`
   */
  docIdPath?: string
}

/**
 * Utiltily Interface so you can have type checked sub functions for your custom parameters
 */
export interface IParametersCustom<T> extends IParametersBase {
  custom?: T
}

/**
 * For when you don't want typechecked custom parameters :)
 */
export interface IParameters extends IParametersBase {
  custom?: any
}

export type getPermitted = (currentType: number, requiredType?: number) => Promise<boolean>

export type validate = (field: IField) => Promise<any>

export interface IErrorBase {
  type: number | string
}
export interface IError extends IErrorBase {
  custom?: any
}

export interface IErrorCustom<T> extends IErrorBase {
  custom?: T
}

export interface IFieldError {
  mapTo: string
  errors: IError[]
}
export interface IDiff {
  mapTo: string
  prev: any
  new: any
}

export interface IFormError {
    model: string
    errors: IFieldError[]
}

export type GetPermissions =
(currentUser: any, options: IParameters) => Promise<boolean | null>

export type GetPermissionCustom<T> =
(currentUser: any, options: IParametersCustom<T>) => Promise<boolean | null>

export interface IPermTypes {
  create?: string
  update?: string
  find?: string
  get?: string
  remove?: string
}

export interface IPermissionsForForm {
  /**
   * Specify a PermissionType as string
   */
  required?: IPermTypes
  /**
   * This will let you override the check by returning a boolean if the user should be allowed.
   * Return null if you wish the xpfw-native check based on `required` permissions to return its own result
   * That way you can limit your `check`-function to e.g. only return on create functions.
   */
  check?: GetPermissions | GetPermissionCustom<any>
  /**
   * Make sure that the result of a `validateForm` call has the current userId set at this key
   *
   */
  requireOwner?: string | IPermTypes
  /**
   * For certain types or custom checks you can pass options here
   * e.g. for the Role-check you can specify which roles need to be presented
   */
  custom?: any
  /**
   * If your users ID is Permissions.Server or Permissions.Admin then allow any action
   * Set to true to disable
   */
  disableAdminOverride?: boolean
}

export interface IUiClient {
  client: any
  connectTo: (url: string, options?: {
    authOptions?: any,
    makeAuth?: boolean,
    useRest?: boolean,
    userStore?: any
    /**
     * Pass the dbStore to get realtime updates
     */
    dbStore?: any
    /**
     * Realtime Updates will only be fetched for collections defined here
     */
    collections?: string[]
  }) => void
  disconnect: () => void,
  login: (loginData: any) => Promise<{user: any, custom?: any}>,
  register: (registerData: any) => Promise<any>,
  logout: () => Promise<any>,
  get: (collection: string, id: any) => Promise<any>,
  remove: (collection: string, id: any) => Promise<any>,
  create: (collection: string, createData: any) => Promise<any>,
  find: (collection: string, createData: any) => Promise<any>,
  patch: (collection: string, id: any, createData: any) => Promise<any>,
}
