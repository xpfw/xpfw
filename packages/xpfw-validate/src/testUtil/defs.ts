import { FieldType, Permission, RequiredType } from "../globals"
import { IField, IForm } from "../typeDef"

const RequiredTextField: IField = {
  mapTo: "myString",
  type: FieldType.Text,
  validate: {required: {type: RequiredType.Always}}
}
const NumberField: IField = {
  mapTo: "myNum",
  type: FieldType.Number,
  validate: {}
}

const FormNumberAndRequiredText: IForm = {
  model: "formModel",
  collection: "simpleTestCol",
  sections: [{fields: [RequiredTextField]}, {fields: [NumberField]}]
}

const TextArrayField: IField = {
  mapTo: "myArray",
  type: FieldType.Array,
  validate: {
    type: FieldType.Text
  }
}

const CreatedAtField: IField = {
  mapTo: "createdAt",
  type: FieldType.Date,
  validate: {
    hide: {create: true, update: false, find: false, remove: false}
  }
}

const FormNumberReqTextWithCreationDate: IForm = {
  model: "formDateModel",
  collection: "DatesimpleTestCol",
  sections: [{fields: [RequiredTextField]}, {fields: [NumberField, CreatedAtField]}],
  options: {
    addCreatedAt: true
  }
}
const FormArray: IForm = {
  model: "arrayModel",
  collection: "arrayCol",
  sections: [{fields: [TextArrayField]}, {fields: [NumberField, CreatedAtField]}],
  options: {
    addCreatedAt: true
  }
}

const RelationshipSingleField: IField = {
  mapTo: "ownedBy",
  type: FieldType.RelationshipSingle,
  validate: {
    relationshipCollection: FormNumberAndRequiredText.collection,
    relationshipNamePath: RequiredTextField.mapTo
  }
}

const RelationshipMultiField: IField = {
  mapTo: "ownedByMulti",
  type: FieldType.RelationshipMulti,
  validate: {
    relationshipCollection: FormNumberAndRequiredText.collection,
    relationshipNamePath: RequiredTextField.mapTo
  }
}

const FormRelationshipToNumAndRequiredText: IForm = {
  model: "relationshipFormModel",
  collection: "myRelationshipCollection",
  sections: [{fields: [RelationshipSingleField, RelationshipMultiField]}, {fields: [NumberField]}]
}

const NumberFieldNested: IField = {
  mapTo: "my.nested.number",
  type: FieldType.Number,
  validate: {}
}

const NumberObjectField: IField = {
  mapTo: "myObj",
  type: FieldType.Object,
  validate: {
    objectDef: [NumberField, NumberFieldNested]
  }
}

const FormForUpdateTest: IForm = {
  model: "updateFormTest",
  sections: [{fields: [NumberFieldNested]}]
}

const NameFieldPublic: IField = {
  mapTo: "guestName",
  type: FieldType.Text,
  validate: {
    permission: {
      find: Permission.Guest,
      update: Permission.Guest,
      create: Permission.Guest,
      remove: Permission.Guest
    }
  }
}

const BooleanFieldWithMultiPermissions: IField = {
  mapTo: "permbool",
  type: FieldType.Boolean,
  validate: {
    permission: {
      find: Permission.User,
      update: Permission.Server,
      create: Permission.User,
      remove: Permission.Server
    }
  }
}

const ObjectWithDifferingPermsPerKey: IField = {
  mapTo: "permObj",
  type: FieldType.Text,
  validate: {
    objectDef: [NameFieldPublic, BooleanFieldWithMultiPermissions]
  }
}

const FormWithPermissionsOnFields: IForm = {
  model: "permFormTest",
  sections: [{fields: [NameFieldPublic, BooleanFieldWithMultiPermissions, ObjectWithDifferingPermsPerKey]}]
}

const FormWithPermissionsOnFieldsAndGeneralPerms: IForm = {
  model: "permFormTest",
  sections: [{fields: [NameFieldPublic, BooleanFieldWithMultiPermissions, ObjectWithDifferingPermsPerKey]}],
  permissions: {
    required: {
      create: Permission.User,
      find: Permission.Guest,
      get: Permission.Guest,
      update: Permission.Owner,
      remove: Permission.Owner
    }
  }
}

const TestDefs = {
  RequiredTextField, NumberField, FormNumberAndRequiredText,
  NumberFieldNested, FormForUpdateTest,  NumberObjectField, RelationshipMultiField, RelationshipSingleField,
  NameFieldPublic, BooleanFieldWithMultiPermissions, ObjectWithDifferingPermsPerKey, FormWithPermissionsOnFields,
  FormWithPermissionsOnFieldsAndGeneralPerms, FormRelationshipToNumAndRequiredText,
  FormNumberReqTextWithCreationDate, CreatedAtField, FormArray, TextArrayField
}

export default TestDefs
export {
  RequiredTextField, NumberField, FormNumberAndRequiredText,
  NumberFieldNested, FormForUpdateTest,  NumberObjectField, RelationshipMultiField, RelationshipSingleField,
  NameFieldPublic, BooleanFieldWithMultiPermissions, ObjectWithDifferingPermsPerKey, FormWithPermissionsOnFields,
  FormWithPermissionsOnFieldsAndGeneralPerms, FormRelationshipToNumAndRequiredText, FormNumberReqTextWithCreationDate,
  CreatedAtField, FormArray, TextArrayField
}
