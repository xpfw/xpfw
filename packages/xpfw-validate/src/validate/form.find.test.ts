import { set } from "lodash"
import { FieldType, Method, RequiredType } from "../globals"
import ValidationRegistry from "../registry"
import { NumberField, NumberFieldNested, NumberObjectField } from "../testUtil/defs"
import { IField, IForm, IParameters } from "../typeDef"
import validateForm from "./form"
import validateQueryObject from "./queryObject"

const form: IForm = {
  model: "sdf",
  sections: [{fields: [NumberField, NumberFieldNested, NumberObjectField]}]
}

const untypedValidateForm: any = validateForm

const params: IParameters = {registry: ValidationRegistry, method: Method.Find}

test("Verify validateFormWithMongoQuery", async () => {
  await expect(untypedValidateForm({myNum: 432}, form)).resolves.toMatchSnapshot("Successfully Validate Number Object")
  await expect(untypedValidateForm({$inc: {myNum: 234}}, form))
  .resolves.toMatchSnapshot("dont parse update operator")

  await expect(untypedValidateForm({myNum: {$gte: 234}}, form, params))
  .resolves.toMatchSnapshot("Correctly parse sameTypeOperator")

  await expect(untypedValidateForm({myNum: {$in: [54, 123, 432, 1543, 132]}}, form, params))
  .resolves.toMatchSnapshot("Correctly parse arrayQueryOperator")

  await expect(untypedValidateForm({myNum: {$gte: "2sd34"}}, form, params))
  .rejects.toMatchSnapshot("Throw error on sameTypeOperator if types do not match")

  await expect(untypedValidateForm({myNum: {$in: [54, 123, 432, 1543, "gsfd"]}}, form, params))
  .rejects.toMatchSnapshot("Throw error on arrayQueryOperator if one of the types does not match")

  await expect(untypedValidateForm({$or: [
    {myNum: {$gte: 234}},
    {myNum: {$in: [54, 123, 432, 1543, 5432]}}
  ]}, form, params))
  .resolves.toMatchSnapshot("Correctly parse arrayedOperators")
  await expect(untypedValidateForm({
    $or: [
      {myNum: {$gte: 234}},
      {myNum: {$in: [54, 123, 432, 1543, 5432]}}
    ],
    $sort: {myNum: -1, the: 53},
    $limit: 20,
    $skip: 10
  }, form, params))
  .resolves.toMatchSnapshot("Correctly parse arrayedOperators with $sort")

  await expect(untypedValidateForm({$or: [
    {myNum: {$gte: 234}},
    {myNum: {$in: [54, 123, {}, 1543, 5432]}}
  ]}, form, params))
  .rejects.toMatchSnapshot("Correctly reject invalid arrayedOperators")

  await expect(untypedValidateForm({$or: "wuff"}, form, params))
  .rejects.toMatchSnapshot("Reject empty special qury")

  await expect(untypedValidateForm({
    $or: [
      {myNum: {$gte: 234}},
      {myNum: {$in: [54, 123, 432, 1543, 5432]}}
    ],
    $sort: {myNum: -1, the: 53, ma: "la", bu: false},
    $limit: 20,
    $skip: 10
  }, form, params))
  .rejects.toMatchSnapshot("Reject faulty $sort")

  await expect(validateQueryObject({}, form, params))
  .resolves.toMatchSnapshot("Resolve null if nothing suiting found")
})
