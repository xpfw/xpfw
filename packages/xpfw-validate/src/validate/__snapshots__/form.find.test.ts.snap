// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify validateFormWithMongoQuery: Correctly parse arrayQueryOperator 1`] = `
Object {
  "myNum": Object {
    "$in": Array [
      54,
      123,
      432,
      1543,
      132,
    ],
  },
}
`;

exports[`Verify validateFormWithMongoQuery: Correctly parse arrayedOperators 1`] = `
Object {
  "$or": Array [
    Object {
      "myNum": Object {
        "$gte": 234,
      },
    },
    Object {
      "myNum": Object {
        "$in": Array [
          54,
          123,
          432,
          1543,
          5432,
        ],
      },
    },
  ],
}
`;

exports[`Verify validateFormWithMongoQuery: Correctly parse arrayedOperators with $sort 1`] = `
Object {
  "$limit": 20,
  "$or": Array [
    Object {
      "myNum": Object {
        "$gte": 234,
      },
    },
    Object {
      "myNum": Object {
        "$in": Array [
          54,
          123,
          432,
          1543,
          5432,
        ],
      },
    },
  ],
  "$skip": 10,
  "$sort": Object {
    "myNum": -1,
  },
}
`;

exports[`Verify validateFormWithMongoQuery: Correctly parse sameTypeOperator 1`] = `
Object {
  "myNum": Object {
    "$gte": 234,
  },
}
`;

exports[`Verify validateFormWithMongoQuery: Correctly reject invalid arrayedOperators 1`] = `
Object {
  "errors": Array [
    Object {
      "type": 3,
    },
  ],
  "mapTo": "any",
}
`;

exports[`Verify validateFormWithMongoQuery: Reject empty special qury 1`] = `"empty query"`;

exports[`Verify validateFormWithMongoQuery: Reject faulty $sort 1`] = `
Object {
  "errors": Array [
    Object {
      "type": 3,
    },
  ],
  "mapTo": "$sort",
}
`;

exports[`Verify validateFormWithMongoQuery: Resolve null if nothing suiting found 1`] = `null`;

exports[`Verify validateFormWithMongoQuery: Successfully Validate Number Object 1`] = `
Object {
  "myNum": 432,
}
`;

exports[`Verify validateFormWithMongoQuery: Throw error on arrayQueryOperator if one of the types does not match 1`] = `
Object {
  "errors": Array [
    Object {
      "type": 3,
    },
  ],
  "mapTo": "any",
}
`;

exports[`Verify validateFormWithMongoQuery: Throw error on sameTypeOperator if types do not match 1`] = `
Object {
  "errors": Array [
    Object {
      "type": 3,
    },
  ],
  "mapTo": "myNum",
}
`;

exports[`Verify validateFormWithMongoQuery: dont parse update operator 1`] = `Object {}`;
